File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(graph) {
    var answer = 0;
  //     최단거리니까 BFS접근
  let dx = [-1, 1, 0, 0];
  let dy = [0, 0, -1, 1];
  let n = graph.length;
  let m = graph[0].length;
  function BFS(x, y) {
    let queue = [];
    queue.push([x, y]);
    while (queue.length) {
      let [x, y] = queue.shift();
      if (x === n - 1 && y === m - 1) return;
      for (let i = 0; i < 4; i++) {
        let nx = x + dx[i];
        let ny = y + dy[i];
        if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
        if (graph[nx][ny] === 1 ) {
          graph[nx][ny] = graph[x][y] + 1;
          queue.push([nx, ny]);
        }
          
      }
        graph[x][y] = 0
    }
  }
  BFS(0, 0);
  answer = graph[n - 1][m - 1];
    if(answer === 1) answer = -1
    return answer;
}

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = 0
    
    dx = [-1,1,0,0]
    dy = [0,0,-1,1]
    visited = [[False for _ in range(len(maps[0]))] for _ in range(len(maps))]
    
    def bfs(graph,visited):
        queue = deque([(0,0)])
        while queue:
            x,y = queue.popleft()
            visited[x][y] = True
            for i in range(4):
                nx = x+dx[i]
                ny = y+dy[i]
                if nx<0 or ny<0 or nx>=len(maps) or ny>=len(maps[0]):
                    continue
                if graph[nx][ny] == 1 and not visited[nx][ny]:
                    queue.append((nx,ny))
                    graph[nx][ny] = graph[x][y] + 1
    bfs(maps,visited)
    if maps[-1][-1] == 1:
        return -1
    else:
        return maps[-1][-1]
    return answer